File: src/rt.c
#include "rt.h"

void ray_at(ray r, double t, vec3 out_pos) {
    glm_vec3_scale(r.dir, t, out_pos);
    glm_vec3_add(r.orig, out_pos, out_pos);
}

static void hit_set_front_face(ray r, hit_record *rec) {
    rec->front_face = glm_vec3_dot(r.dir, rec->norm) < 0.0;
    if (!rec->front_face)
        glm_vec3_negate(rec->norm);
}

static bool sphere_hit(ray r, interval t, hit_record *out, void *self) {
    sphere *s = self;
    vec3 oc;
    glm_vec3_sub(s->center, r.orig, oc);
    double a = glm_vec3_norm2(r.dir);
    double h = glm_vec3_dot(r.dir, oc);
    double c = glm_vec3_norm2(oc) - s->radius * s->radius;
    double discriminant = h * h - a * c;
    if (discriminant < 0.0)
        return false;
    double d_sqrt = sqrt(discriminant);
    double root = (h - d_sqrt) / a;
    if (!interval_sorrounds(t, root)) {
        root = (h + d_sqrt) / a;
        if (!interval_sorrounds(t, root))
            return false;
    }

    out->mat = s->mat;
    ray_at(r, root, out->p);
    glm_vec3_sub(out->p, s->center, out->norm);
    glm_vec3_scale(out->norm, 1.0 / s->radius, out->norm);
    out->t = root;
    hit_set_front_face(r, out);

    return true;
}

hittable sphere_hittable(sphere *s) {
    return (hittable){
        .hit = sphere_hit,
        .self = s,
    };
}

static bool hittable_list_hit(ray r, interval t, hit_record *out, void *self) {
    hittable_list *s = self;
    hit_record tmp_rec;
    bool hit_anything = false;
    double closest = t.max;

    for (size_t i = 0; i < s->len; ++i) {
        if (s->objects[i].hit(r, (interval){t.min, closest}, &tmp_rec,
                              s->objects[i].self)) {
            hit_anything = true;
            closest = tmp_rec.t;
            *out = tmp_rec;
        }
    }

    return hit_anything;
}

hittable hittable_list_hittable(hittable_list *h) {
    return (hittable){
        .hit = hittable_list_hit,
        .self = h,
    };
}

static bool lambertian_scatter(ray r, hit_record h, color out_attenuation,
                               ray *out_scattered, void *self) {
    lambertian *s = self;
    random_normalized_vec3(out_scattered->dir);
    glm_vec3_add(r.dir, h.norm, r.dir);
    if (vec3_is_near_zero(r.dir))
        glm_vec3_copy(h.norm, r.dir);
    glm_vec3_copy(h.p, out_scattered->orig);
    glm_vec3_copy(s->albedo, out_attenuation);
    return true;
}

material lambertian_material(lambertian *l) {
    return (material){
        .scatter = lambertian_scatter,
        .self = l,
    };
}

static bool metal_scatter(ray r, hit_record h, color out_attenuation,
                          ray *out_scattered, void *self) {
    metal *s = self;
    glm_vec3_reflect(r.dir, h.norm, out_scattered->dir);
    vec3 fuzz;
    random_normalized_vec3(fuzz);
    glm_vec3_scale(fuzz, s->fuzziness, fuzz);
    glm_vec3_add(out_scattered->dir, fuzz, out_scattered->dir);
    glm_vec3_copy(h.p, out_scattered->orig);
    glm_vec3_copy(s->albedo, out_attenuation);
    return true;
}

material metal_material(metal *m) {
    return (material){
        .scatter = metal_scatter,
        .self = m,
    };
}

static double reflectance(double cosine, double refraction_idx) {
    // Schlick's approximation for reflectance
    double r0 = (1 - refraction_idx) / (1 + refraction_idx);
    r0 = r0 * r0;
    return r0 + (1 - r0) * pow(1 - cosine, 5);
}

static bool dielectric_scatter(ray r, hit_record h, color out_attenuation,
                               ray *out_scattered, void *self) {
    dielectric *s = self;
    glm_vec3_copy((color){1.0, 1.0, 1.0}, out_attenuation);
    double refraction_idx =
        h.front_face ? 1.0 / s->refraction_idx : s->refraction_idx;
    glm_vec3_normalize(r.dir);

    vec3 tmp;
    glm_vec3_negate_to(r.dir, tmp);
    double cos_theta = fmin(glm_vec3_dot(tmp, h.norm), 1.0);
    double sin_theta = sqrt(1.0 - cos_theta * cos_theta);
    bool cannot_refract = refraction_idx * sin_theta > 1.0;
    if (cannot_refract ||
        reflectance(cos_theta, refraction_idx) > random_double())
        glm_vec3_reflect(r.dir, h.norm, out_scattered->dir);
    else
        glm_vec3_refract(r.dir, h.norm, refraction_idx, out_scattered->dir);

    glm_vec3_copy(h.p, out_scattered->orig);
    return true;
}

material dielectric_material(dielectric *d) {
    return (material){
        .scatter = dielectric_scatter,
        .self = d,
    };
}

static void ray_color(ray r, int depth, hittable world, color out) {
    if (depth <= 0) {
        glm_vec3_copy((color){0.0, 0.0, 0.0}, out);
        return;
    }
    hit_record h;
    if (world.hit(r, (interval){0.001, infinity}, &h, world.self)) {
        ray scattered;
        color attenuation;
        if (h.mat.scatter(r, h, attenuation, &scattered, h.mat.self)) {
            ray_color(scattered, depth - 1, world, out);
            glm_vec3_mul(out, attenuation, out);
            return;
        }
        glm_vec3_copy((color){0.0, 0.0, 0.0}, out);
        return;
    }

    glm_vec3_normalize(r.dir);
    double a = 0.5 * (r.dir[1] + 1.0);
    color start = {1.0, 1.0, 1.0};
    color end = {0.5, 0.7, 1.0};
    glm_vec3_scale(start, 1.0 - a, start);
    glm_vec3_scale(end, a, end);
    glm_vec3_add(start, end, out);
}

static double linear_to_gamma(double linear) {
    if (linear > 0)
        return sqrt(linear);
    return 0.0;
}

static void write_color(int width, int x, int y, const color color,
                        pixel out_buf[]) {
    interval i = {0.0, 0.999};
    pixel p;
    p.r = 256.0 * interval_clamp(i, linear_to_gamma(color[0]));
    p.g = 256.0 * interval_clamp(i, linear_to_gamma(color[1]));
    p.b = 256.0 * interval_clamp(i, linear_to_gamma(color[2]));
    out_buf[width * y + x] = p;
}

static void random_square_offset(vec3 out) {
    out[0] = random_double() - 0.5;
    out[1] = random_double() - 0.5;
    out[2] = 0.0;
}

static ray camera_get_ray(camera cam, int i, int j) {
    ray ret;
    vec3 offset;
    random_square_offset(offset);
    glm_vec3_copy(cam.center, ret.orig);
    if (cam.defocus_angle > 0.0) {
        vec3 tmp;
        random_in_unit_disc_vec3(tmp);
        glm_vec3_muladds(cam.defocus_disk_u, tmp[0], ret.orig);
        glm_vec3_muladds(cam.defocus_disk_v, tmp[1], ret.orig);
    }

    vec3 du;
    glm_vec3_scale(cam.pixel_du, j + offset[0], du);
    vec3 dv;
    glm_vec3_scale(cam.pixel_dv, i + offset[1], dv);
    vec3 px_center;
    glm_vec3_add(cam.pixel00, du, px_center);
    glm_vec3_add(px_center, dv, px_center);
    glm_vec3_sub(px_center, ret.orig, ret.dir);
    return ret;
}

void camera_render(camera cam, hittable world, pixel out_buf[]) {
    // printf("P3\n%d %d\n255\n", cam.image_width, cam.image_height);
    for (int i = 0; i < cam.image_height; ++i) {
        fprintf(stderr, "\rScanlines remaining: %5d", cam.image_height - i);
        fflush(stderr);
        for (int j = 0; j < cam.image_width; ++j) {
            color color = {0.0, 0.0, 0.0};
            for (int sample = 0; sample < cam.samples_per_px; ++sample) {
                vec3 sample_color;
                ray ray = camera_get_ray(cam, i, j);
                ray_color(ray, cam.max_depth, world, sample_color);
                glm_vec3_add(color, sample_color, color);
            }
            glm_vec3_scale(color, 1.0 / cam.samples_per_px, color);
            write_color(cam.image_width, j, i, color, out_buf);
        }
    }
    fprintf(stderr, "\rDone                              \n");
}
camera camera_init(camera_desc d) {
    camera ret;
    vec3 tmp;
    ret.image_width = d.image_width;
    ret.samples_per_px = d.samples_per_px;
    ret.max_depth = d.max_depth;
    ret.defocus_angle = d.defocus_angle;
    ret.image_height = maxi(1, ret.image_width / d.aspect_ratio);
    glm_vec3_copy(d.lookfrom, ret.center);

    double h = tan(d.fovy / 2);
    double viewport_height = 2.0 * h * d.focus_dist;
    double viewport_width =
        viewport_height * (double)ret.image_width / ret.image_height;

    vec3 u, v, w, vup;
    glm_vec3_copy((vec3){0.0, 1.0, 0.0}, vup);
    glm_vec3_sub(d.lookfrom, d.lookat, w);
    glm_normalize(w);
    glm_vec3_cross(vup, w, u);
    glm_vec3_normalize(u);
    glm_vec3_cross(w, u, v);

    vec3 viewport_u, viewport_v;
    glm_vec3_scale(u, viewport_width, viewport_u);
    glm_vec3_scale(v, viewport_height, viewport_v);
    glm_vec3_negate(viewport_v);
    glm_vec3_scale(viewport_u, 1.0 / ret.image_width, ret.pixel_du);
    glm_vec3_scale(viewport_v, 1.0 / ret.image_height, ret.pixel_dv);

    vec3 viewport_upper_left;
    glm_vec3_copy(ret.center, viewport_upper_left);
    glm_vec3_scale(w, d.focus_dist, tmp);
    glm_vec3_sub(viewport_upper_left, tmp, viewport_upper_left);
    glm_vec3_scale(viewport_u, 0.5, tmp);
    glm_vec3_sub(viewport_upper_left, tmp, viewport_upper_left);
    glm_vec3_scale(viewport_v, 0.5, tmp);
    glm_vec3_sub(viewport_upper_left, tmp, viewport_upper_left);

    glm_vec3_add(ret.pixel_du, ret.pixel_dv, tmp);
    glm_vec3_scale(tmp, 0.5, tmp);
    glm_vec3_add(viewport_upper_left, tmp, ret.pixel00);

    double defocus_radius = d.focus_dist * tan(d.defocus_angle / 2.0);
    glm_vec3_scale(u, defocus_radius, ret.defocus_disk_u);
    glm_vec3_scale(v, defocus_radius, ret.defocus_disk_v);
    return ret;
}

File: src/rt.rs
use crate::rng::fast_random;
use glam::{UVec2, Vec2, Vec3};
use image::Rgb32FImage;
use std::f32::{self, consts::PI};
use std::{ops::Index, sync::Arc};

pub const INFINITY: f32 = std::f32::MAX;

#[derive(Debug, Clone, Copy)]
pub struct Interval {
    pub min: f32,
    pub max: f32,
}

impl Interval {
    pub const EMPTY: Self = Self {
        min: INFINITY,
        max: -INFINITY,
    };
    pub const UNIVERSE: Self = Self {
        min: -INFINITY,
        max: INFINITY,
    };

    pub fn new(min: f32, max: f32) -> Self {
        Self { min, max }
    }

    pub fn from_intervals(a: Self, b: Self) -> Self {
        let min = a.min.min(b.min);
        let max = a.max.max(b.max);
        Self { min, max }
    }

    pub fn size(&self) -> f32 {
        self.max - self.min
    }

    pub fn contains(&self, x: f32) -> bool {
        self.min <= x && x <= self.max
    }
    pub fn surrounds(&self, x: f32) -> bool {
        self.min < x && x < self.max
    }

    pub fn expand(&self, delta: f32) -> Self {
        let padding = delta / 2.0;
        Interval {
            min: self.min - padding,
            max: self.max + padding,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Aabb {
    pub x: Interval,
    pub y: Interval,
    pub z: Interval,
}

impl Aabb {
    pub const EMPTY: Self = Self::new(Interval::EMPTY, Interval::EMPTY, Interval::EMPTY);

    pub const fn new(x: Interval, y: Interval, z: Interval) -> Self {
        Self { x, y, z }
    }

    pub fn from_points(a: Vec3, b: Vec3) -> Self {
        let x = if a[0] < b[0] {
            Interval::new(a[0], b[0])
        } else {
            Interval::new(b[0], a[0])
        };
        let y = if a[1] < b[1] {
            Interval::new(a[1], b[1])
        } else {
            Interval::new(b[1], a[1])
        };
        let z = if a[2] < b[2] {
            Interval::new(a[2], b[2])
        } else {
            Interval::new(b[2], a[2])
        };
        Self { x, y, z }
    }

    pub fn from_aabbs(a: Self, b: Self) -> Self {
        Self {
            x: Interval::from_intervals(a.x, b.x),
            y: Interval::from_intervals(a.y, b.y),
            z: Interval::from_intervals(a.z, b.z),
        }
    }

    pub fn longest_axis(&self) -> usize {
        if self.x.size() > self.y.size() {
            if self.x.size() > self.z.size() { 0 } else { 2 }
        } else if self.y.size() > self.z.size() {
            1
        } else {
            0
        }
    }

    pub fn hit(&self, ray: &Ray, mut valid_t: Interval) -> bool {
        for axis in 0..3 {
            let interval = [&self.x, &self.y, &self.z][axis];
            let dir_inv = 1.0 / ray.direction[axis];
            let t0 = (interval.min - ray.origin[axis]) * dir_inv;
            let t1 = (interval.max - ray.origin[axis]) * dir_inv;
            if t0 < t1 {
                valid_t.min = valid_t.min.max(t0);
                valid_t.max = valid_t.max.min(t1);
            } else {
                valid_t.min = valid_t.min.max(t1);
                valid_t.max = valid_t.max.min(t0);
            }

            if valid_t.max <= valid_t.min {
                return false;
            }
        }
        true
    }
}

impl Index<usize> for Aabb {
    type Output = Interval;

    fn index(&self, index: usize) -> &Self::Output {
        match index {
            0 => &self.x,
            1 => &self.y,
            2 => &self.z,
            _ => panic!("{index} is out of range for Aabb"),
        }
    }
}

pub fn random_unit_vec3() -> Vec3 {
    loop {
        let v = 2.0 * fast_random::<Vec3>() - 1.0;
        let len_sq = v.length_squared();
        if len_sq < 1.0
            && let Some(v) = v.try_normalize()
        {
            return v;
        }
    }
}

pub fn random_unit_disk_vec3() -> Vec3 {
    loop {
        let v = 2.0 * fast_random::<Vec3>() - 1.0;
        if v.length_squared() < 1.0 {
            return v;
        }
    }
}

#[derive(Debug, Clone)]
pub struct Ray {
    pub origin: Vec3,
    pub direction: Vec3,
    pub time: f32,
}

impl Ray {
    pub fn at(&self, t: f32) -> Vec3 {
        self.origin + self.direction * t
    }
}

#[derive(Clone)]
pub struct HitRecord<'material> {
    pub point: Vec3,
    pub normal: Vec3,
    pub tex_coords: Vec2,
    pub material: &'material dyn Material,
    pub t: f32,
    pub front_face: bool,
}

pub trait Hittable: Sync + Send {
    fn hit(&self, ray: &Ray, valid_t: Interval) -> Option<HitRecord<'_>>;
    fn bounding_box(&self) -> Aabb;
}

pub trait Material: Sync + Send {
    fn scatter(&self, ray: &Ray, record: &HitRecord) -> Option<(Vec3, Ray)>;
}

#[derive(Debug, Clone)]
pub struct Camera {
    image_width: u32,
    image_height: u32,

    center: Vec3,
    pixel00: Vec3,
    pixel_du: Vec3,
    pixel_dv: Vec3,

    defocus_angle: f32,
    defocus_disk_u: Vec3,
    defocus_disk_v: Vec3,

    samples_per_px: u32,
    max_depth: u32,
}

#[derive(Debug, Clone)]
pub struct CameraDesc {
    pub image_width: u32,
    pub aspect_ratio: f32,

    pub look_from: Vec3,
    pub look_at: Vec3,

    pub defocus_angle: f32,
    pub focus_dist: f32,

    pub fovy: f32,

    pub samples_per_px: u32,
    pub max_depth: u32,
}

impl Camera {
    pub fn new(desc: &CameraDesc) -> Self {
        let image_height = (desc.image_width as f32 / desc.aspect_ratio) as u32;
        let h = (desc.fovy / 2.0).tan();

        let viewport_height = 2.0 * h * desc.focus_dist;
        let viewport_width = viewport_height * desc.image_width as f32 / image_height as f32;

        let w = (desc.look_from - desc.look_at).normalize();
        let u = Vec3::Y.cross(w).normalize();
        let v = w.cross(u);

        let viewport_u = viewport_width * u;
        let viewport_v = viewport_height * -v;
        let pixel_du = viewport_u / (desc.image_width as f32);
        let pixel_dv = viewport_v / (image_height as f32);

        let viewport_top_left =
            desc.look_from - (desc.focus_dist * w) - viewport_u / 2.0 - viewport_v / 2.0;
        let pixel00 = viewport_top_left + 0.5 * (pixel_du + pixel_dv);

        let defocus_radius = desc.focus_dist * (desc.defocus_angle / 2.0).tan();
        let defocus_disk_u = u * defocus_radius;
        let defocus_disk_v = v * defocus_radius;
        Self {
            image_width: desc.image_width,
            image_height,
            center: desc.look_from,
            pixel00,
            pixel_du,
            pixel_dv,

            defocus_angle: desc.defocus_angle,
            defocus_disk_u,
            defocus_disk_v,

            samples_per_px: desc.samples_per_px,
            max_depth: desc.max_depth,
        }
    }

    pub fn width(&self) -> u32 {
        self.image_width
    }
    pub fn height(&self) -> u32 {
        self.image_height
    }

    fn get_ray(&self, coords: UVec2) -> Ray {
        let x = coords.x as f32 + fast_random::<f32>() - 0.5;
        let y = coords.y as f32 + fast_random::<f32>() - 0.5;
        let px_loc = self.pixel00 + x * self.pixel_du + y * self.pixel_dv;
        let origin = if self.defocus_angle > 0.0 {
            let offset = random_unit_disk_vec3();
            self.center + offset.x * self.defocus_disk_u + offset.y * self.defocus_disk_v
        } else {
            self.center
        };
        let time = fast_random();
        Ray {
            origin,
            direction: px_loc - origin,
            time,
        }
    }
}

fn ray_color(ray: &Ray, depth: u32, world: &dyn Hittable) -> Vec3 {
    if depth == 0 {
        return Vec3::ZERO;
    }

    let pos_interval = Interval {
        min: 0.001,
        max: INFINITY,
    };
    if let Some(record) = world.hit(ray, pos_interval) {
        if let Some((attenuation, scattered)) = record.material.scatter(ray, &record) {
            return attenuation * ray_color(&scattered, depth - 1, world);
        }
        return Vec3::X;
    }

    let dir = ray.direction.normalize();
    let a = 0.5 * (dir.y + 1.0);
    (1.0 - a) * Vec3::new(1.0, 1.0, 1.0) + a * Vec3::new(0.5, 0.7, 1.0)
}

pub fn render_pixel(camera: &Camera, world: &dyn Hittable, coords: UVec2) -> Vec3 {
    let mut color = Vec3::ZERO;
    for _ in 0..camera.samples_per_px {
        color += ray_color(&camera.get_ray(coords), camera.max_depth, world);
    }
    color / (camera.samples_per_px as f32)
}

#[derive(Clone)]
pub struct Sphere {
    pub center: Ray,
    pub radius: f32,
    pub material: Arc<dyn Material>,
}

impl Sphere {
    pub fn stationary(center: Vec3, radius: f32, material: Arc<dyn Material>) -> Self {
        Self {
            center: Ray {
                origin: center,
                direction: Vec3::ZERO,
                time: 0.0,
            },
            radius,
            material,
        }
    }

    pub fn moving(center1: Vec3, center2: Vec3, radius: f32, material: Arc<dyn Material>) -> Self {
        Self {
            center: Ray {
                origin: center1,
                direction: center2 - center1,
                time: 0.0,
            },
            radius,
            material,
        }
    }
}

impl Hittable for Sphere {
    fn hit(&self, ray: &Ray, valid_t: Interval) -> Option<HitRecord<'_>> {
        let center = self.center.at(ray.time);
        let oc = center - ray.origin;
        let a = ray.direction.length_squared();
        let h = ray.direction.dot(oc);
        let c = oc.length_squared() - self.radius * self.radius;
        let discriminant = h * h - a * c;
        if discriminant < 0.0 {
            return None;
        }

        let discriminant_sqrt = discriminant.sqrt();
        let mut root = (h - discriminant_sqrt) / a;
        if !valid_t.surrounds(root) {
            root = (h + discriminant_sqrt) / a;
            if !valid_t.surrounds(root) {
                return None;
            }
        }

        let point = ray.at(root);
        let outward_normal = (point - center) / self.radius;
        let front_face = ray.direction.dot(outward_normal) < 0.0;
        let normal = if front_face {
            outward_normal
        } else {
            -outward_normal
        };
        let theta = f32::acos(-outward_normal.y);
        let phi = f32::atan2(-outward_normal.z, outward_normal.x) + PI;
        let u = phi / (2.0 * PI);
        let v = theta / PI;
        let tex_coords = Vec2::new(u, v);

        Some(HitRecord {
            point,
            normal,
            tex_coords,
            t: root,
            material: &*self.material,
            front_face,
        })
    }

    fn bounding_box(&self) -> Aabb {
        let rvec = Vec3::splat(self.radius);
        let box0 = Aabb::from_points(self.center.at(0.0) - rvec, self.center.at(0.0) + rvec);
        let box1 = Aabb::from_points(self.center.at(1.0) - rvec, self.center.at(1.0) + rvec);
        Aabb::from_aabbs(box0, box1)
    }
}

impl Hittable for Vec<Box<dyn Hittable>> {
    fn hit(&self, ray: &Ray, valid_t: Interval) -> Option<HitRecord<'_>> {
        let mut closest_hit = None;
        let mut closest_interval = valid_t;
        for hittable in self {
            if let Some(record) = hittable.hit(ray, closest_interval) {
                closest_interval.max = record.t;
                closest_hit = Some(record);
            }
        }
        closest_hit
    }

    fn bounding_box(&self) -> Aabb {
        let mut bbox = Aabb::EMPTY;
        for hittable in self {
            bbox = Aabb::from_aabbs(bbox, hittable.bounding_box());
        }
        bbox
    }
}

type BvhLink = Arc<dyn Hittable>;
#[derive(Clone)]
pub struct BvhNode {
    left: BvhLink,
    right: BvhLink,
    bbox: Aabb,
}

impl BvhNode {
    pub fn create(hittables: &mut [Arc<dyn Hittable>]) -> Self {
        if hittables.len() == 0 {
            panic!("must create BvhNode with at least one hittable");
        }

        let mut bbox = Aabb::EMPTY;
        for hittable in &mut *hittables {
            bbox = Aabb::from_aabbs(bbox, hittable.bounding_box());
        }
        if hittables.len() == 1 {
            return BvhNode {
                left: hittables[0].clone(),
                right: hittables[0].clone(),
                bbox,
            };
        }
        if hittables.len() == 2 {
            return BvhNode {
                left: hittables[0].clone(),
                right: hittables[1].clone(),
                bbox,
            };
        }

        let axis = bbox.longest_axis();
        hittables.sort_unstable_by(|a, b| {
            a.bounding_box()[axis]
                .min
                .total_cmp(&b.bounding_box()[axis].min)
        });
        let mid = hittables.len() / 2;
        let (left, right) = hittables.split_at_mut(mid);
        let left = Arc::new(BvhNode::create(left));
        let right = Arc::new(BvhNode::create(right));
        BvhNode { left, right, bbox }
    }
}

impl Hittable for BvhNode {
    fn hit(&self, ray: &Ray, valid_t: Interval) -> Option<HitRecord<'_>> {
        if !self.bbox.hit(ray, valid_t) {
            return None;
        }

        let left = self.left.hit(ray, valid_t);
        let max_t = if let Some(ref record) = left {
            record.t
        } else {
            valid_t.max
        };
        let right = self.right.hit(ray, Interval::new(valid_t.min, max_t));

        right.or(left)
    }

    fn bounding_box(&self) -> Aabb {
        self.bbox
    }
}

#[derive(Clone)]
pub struct Lambertian {
    pub texture: Arc<dyn Texture>,
}

impl Lambertian {
    pub fn from_color(albedo: Vec3) -> Self {
        Self {
            texture: Arc::new(ColorTexture { albedo }),
        }
    }
}

impl Material for Lambertian {
    fn scatter(&self, ray: &Ray, record: &HitRecord) -> Option<(Vec3, Ray)> {
        let mut out_dir = random_unit_vec3() + record.normal;
        if out_dir.length_squared() < 1e-4 {
            out_dir = record.normal;
        }
        Some((
            self.texture.value(record.tex_coords, record.point),
            Ray {
                direction: out_dir,
                origin: record.point,
                time: ray.time,
            },
        ))
    }
}

#[derive(Debug, Clone)]
pub struct Metal {
    pub albedo: Vec3,
    pub fuzziness: f32,
}

impl Material for Metal {
    fn scatter(&self, ray: &Ray, record: &HitRecord) -> Option<(Vec3, Ray)> {
        Some((
            self.albedo,
            Ray {
                origin: record.point,
                direction: ray.direction.reflect(record.normal)
                    + self.fuzziness * random_unit_vec3(),
                time: ray.time,
            },
        ))
    }
}

#[derive(Debug, Clone)]
pub struct Dielectric {
    pub refraction_index: f32,
}

impl Dielectric {
    fn reflectance(&self, cos_theta: f32) -> f32 {
        // Schlick's approximation for reflectance
        let r0 = (1.0 - self.refraction_index) / (1.0 + self.refraction_index);
        let r0 = r0 * r0;
        r0 + (1.0 - r0) * (1.0 - cos_theta).powi(5)
    }
}

impl Material for Dielectric {
    fn scatter(&self, ray: &Ray, record: &HitRecord) -> Option<(Vec3, Ray)> {
        let refraction_index = if record.front_face {
            1.0 / self.refraction_index
        } else {
            self.refraction_index
        };
        let unit_direction = ray.direction.normalize();
        let cos_theta = (-unit_direction).dot(record.normal).min(1.0);
        let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();
        let cannot_refract = refraction_index * sin_theta > 1.0;
        let direction = if cannot_refract || self.reflectance(cos_theta) > fast_random() {
            unit_direction.reflect(record.normal)
        } else {
            unit_direction.refract(record.normal, refraction_index)
        };

        Some((
            Vec3::ONE,
            Ray {
                origin: record.point,
                direction,
                time: ray.time,
            },
        ))
    }
}

pub trait Texture: Send + Sync {
    fn value(&self, tex_coords: Vec2, point: Vec3) -> Vec3;
}

#[derive(Debug, Clone)]
pub struct ColorTexture {
    pub albedo: Vec3,
}

impl Texture for ColorTexture {
    fn value(&self, _tex_coords: Vec2, _point: Vec3) -> Vec3 {
        self.albedo
    }
}

#[derive(Clone)]
pub struct CheckerTexture {
    pub scale: f32,
    pub even: Arc<dyn Texture>,
    pub odd: Arc<dyn Texture>,
}

impl Texture for CheckerTexture {
    fn value(&self, tex_coords: Vec2, point: Vec3) -> Vec3 {
        let point_int = (point / self.scale).floor().as_ivec3();
        let is_even = point_int.element_sum() % 2 == 0;
        if is_even {
            self.even.value(tex_coords, point)
        } else {
            self.odd.value(tex_coords, point)
        }
    }
}

impl Texture for Rgb32FImage {
    fn value(&self, tex_coords: Vec2, _point: Vec3) -> Vec3 {
        let u = tex_coords.x.clamp(0.0, 1.0);
        let v = 1.0 - tex_coords.y.clamp(0.0, 1.0);
        let x = (u * self.width() as f32) as u32;
        let y = (v * self.height() as f32) as u32;
        let px: Vec3 = self.get_pixel(x, y).0.into();
        px.powf(2.0)
    }
}
